core/services/ocr2/delegate.go

```diff
@@ -685,57 +680,52 @@ func (d *Delegate) ServicesForSpec(jb job.Job) ([]job.ServiceCtx, error) {
+			Database:                     ocrDB,
+			LocalConfig:                  lc,
+			Logger:                       ocrLogger,
+			MonitoringEndpoint:           d.monitoringEndpointGen.GenMonitoringEndpoint(spec.ContractID, synchronization.OCR2Functions),
+			OffchainConfigDigester:       functionsProvider.OffchainConfigDigester(),
+			OffchainKeyring:              kb,
+			OnchainKeyring:               kb,
+			ReportingPluginFactory:       nil, // To be set by NewFunctionsServices
+		}
 
-	pluginServices, err := pluginOracle.GetServices()
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to get plugin services")
-	}
+		functionsServicesConfig := functions.FunctionsServicesConfig{
+			Job:            jb,
+			PipelineRunner: d.pipelineRunner,
+			JobORM:         d.jobORM,
+			OCR2JobConfig:  d.cfg,
+			DB:             d.db,
+			Chain:          chain,
+			ContractID:     spec.ContractID,
+			Lggr:           lggr,
+			MailMon:        d.mailMon,
+		}
 
-	oracle, err := libocr2.NewOracle(libocr2.OracleArgs{
-		BinaryNetworkEndpointFactory: peerWrapper.Peer2,
-		V2Bootstrappers:              bootstrapPeers,
-		ContractTransmitter:          ocr2Provider.ContractTransmitter(),
-		ContractConfigTracker:        ocr2Provider.ContractConfigTracker(),
-		Database:                     ocrDB,
-		LocalConfig:                  lc,
-		Logger:                       ocrLogger,
-		MonitoringEndpoint:           d.monitoringEndpointGen.GenMonitoringEndpoint(spec.ContractID, synchronization.OCR2Functions),
-		OffchainConfigDigester:       ocr2Provider.OffchainConfigDigester(),
-		OffchainKeyring:              kb,
-		OnchainKeyring:               kb,
-		ReportingPluginFactory:       pluginFactory,
-	})
-	if err != nil {
-		return nil, errors.Wrap(err, "error calling NewOracle")
-	}
+		functionsServices, err := functions.NewFunctionsServices(&sharedOracleArgs, &functionsServicesConfig)
+		if err != nil {
+			return nil, errors.Wrap(err, "error calling NewFunctionsServices")
+		}
 
-	// RunResultSaver needs to be started first, so it's available
-	// to read odb writes. It is stopped last after the OraclePlugin is shut down
-	// so no further runs are enqueued, and we can drain the queue.
-	runResultSaver := ocrcommon.NewResultRunSaver(
-		runResults,
-		d.pipelineRunner,
-		make(chan struct{}),
-		lggr,
-		d.cfg.JobPipelineMaxSuccessfulRuns(),
-	)
+		// RunResultSaver needs to be started first, so it's available
+		// to read odb writes. It is stopped last after the OraclePlugin is shut down
+		// so no further runs are enqueued, and we can drain the queue.
+		runResultSaver := ocrcommon.NewResultRunSaver(
+			runResults,
+			d.pipelineRunner,
+			make(chan struct{}),
+			lggr,
+			d.cfg.JobPipelineMaxSuccessfulRuns(),
+		)
 
-	oracleCtx := job.NewServiceAdapter(oracle)
-	return append([]job.ServiceCtx{runResultSaver, ocr2Provider, oracleCtx}, pluginServices...), nil
+		return append([]job.ServiceCtx{runResultSaver, functionsProvider}, functionsServices...), nil
```

Comment:
So, eventually functions.NewFunctionsServices will also create a Threshold OCR plugin and because we save all the Functions services in this slice, the Threshold plugin will be able to communicate with the other Functions services like the listener.

Reply:
Furthermore, because we are using `append` with destructuring, we can use an arbitrary number of additional services w/o modifying `delegate.go`

Reply:
Yes, we can return an arbitrary number of services (either of type libocr2.Oracle (that run OCR2 plugin) or our own, like the Listener). The communication across those services is something that we need to take care of ourselves. For example when you later create the Queueing Service, you'd need to pass it to both the new OCR2 plugin and to the existing Listener service as well.

On startup, the Node creates one Delegate object for each possible job type. Then for each job spec added to the node, it will call ServicesForSpec() to create all services needed to run an instance of that job type.

The OraclePlugin interface was created a while ago to standardize how all plugins defined services (https://github.com/smartcontractkit/chainlink/pull/6055). But it was too rigid - in particular it allowed only for a single libocr2.Oracle (i.e. a single reporting plugin). From what I see, after our refactor, no other plugin is using that interface. I guess we could we remove it now (@RensR ?).